examples/authentication/authenticate_in_web_application.py

```diff
@@ -12,60 +12,186 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""This example creates an OAuth 2.0 refresh token for the Google Ads API.
+"""This example creates an OAuth2 refresh token using the Web application flow.
 
-This illustrates how to step through the OAuth 2.0 native / installed
-application flow.
+To retrieve the necessary client_secrets JSON file, first generate an OAuth 2.0
+client ID of type 'Web application' in the Google Cloud Console:
 
-It is intended to be run from the command line and requires user input.
+https://console.cloud.google.com
+
+Make sure 'http://localhost:8080' is included the list of authorized redirect
+URIs for this client ID.
+
+Once complete, download the credentials and save the file path so it can be
+passed into this example.
+
+This example is a very simple implementation, for a more detailed example see:
+https://developers.google.com/identity/protocols/oauth2/web-server#python
 """
 
 
 import argparse
+import hashlib
+import os
+import re
+import socket
+import sys
 
-from google_auth_oauthlib.flow import InstalledAppFlow
+from google_auth_oauthlib.flow import Flow
 
 
-SCOPE = "https://www.googleapis.com/auth/adwords"
+_SCOPE = "https://www.googleapis.com/auth/adwords"
+_REDIRECT_URI = "http://localhost:8080"
 
 
 def main(client_secrets_path, scopes):
-    flow = InstalledAppFlow.from_client_secrets_file(
-        client_secrets_path, scopes=scopes
+    """The main method, starts a basic server and initializes an auth request.
+
+    Args:
+        client_secrets_path: a path to where the client secrets JSON file is
+            located on the machine running this example.
+        scopes: a list of API scopes to include in the auth request, see:
+            https://developers.google.com/identity/protocols/oauth2/scopes
+    """
+    flow = Flow.from_client_secrets_file(client_secrets_path, scopes=scopes)
+    flow.redirect_uri = _REDIRECT_URI
+
+    # Create an anti-forgery state token as described here:
+    # https://developers.google.com/identity/protocols/OpenIDConnect#createxsrftoken
+    passthrough_val = hashlib.sha256(os.urandom(1024)).hexdigest()
+
+    authorization_url, state = flow.authorization_url(
+        access_type='offline',
+        state=passthrough_val,
+        include_granted_scopes='true'
+    )
+
+    # Prints the authorization URL so you can paste into your browser. In a
+    # typical web application you would redirect the user to this URL, and they
+    # would be redirected back to "redirect_url" provided earlier after
+    # granting permission.
+    print(authorization_url)
```

Comment:
should be some description in the print output, ie. "Paste this url in your browser:"

and then

print("Waiting for authorization and callback on uri {_REDIRECT_URI}")

Reply:
Done